### PR TITLE
Beta: Attachments

### PR DESCRIPTION
**Pull request by marcbouchenoire**
This PR adds support for attachments to Comments/Notifications.

It's a long-lived branch (a continuation of https://github.com/liveblocks/liveblocks/pull/1888) that has been reviewed in some ways before (general API design and docs mostly), but not the actual implementation yet.

### CODE REVIEW

File: packages/liveblocks-react-ui/src/components/internal/Attachment.tsx [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+function splitFileName(name: string) {
+  const match = name.match(fileExtensionRegex);
+
+  return { base: match?.[1] ?? name, extension: match?.[2] };
+}
+
+function getFileAttachmentIconGlyph(mimeType: string) {
+  if (
+    mimeType === "application/zip" ||
+    mimeType === "application/x-rar-compressed" ||
```
Comment (CTNicholas) [inline packages/liveblocks-react-ui/src/components/internal/Attachment.tsx:48]:
```suggestion
    mimeType === "application/x-rar-compressed" ||
    mimeType === "application/vnd.rar" ||
    mimeType === "application/x-bzip" ||
    mimeType === "application/x-bzip2" ||
```

Listed here as the rar and bzip mimetypes: https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Common_types

File: packages/liveblocks-react-ui/src/components/internal/Attachment.tsx [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+    mimeType === "application/x-zip-compressed" ||
+    mimeType === "application/x-tar" ||
+    mimeType === "application/gzip"
+  ) {
+    return (
+      <path d="M13 15h2v1h-1.5a.5.5 0 0 0 0 1H15v1h-1.5a.5.5 0 0 0 0 1H15v1h-1.5a.5.5 0 0 0 0 1h1a.5.5 0 0 0 .5-.5V20h1.5a.5.5 0 0 0 0-1H15v-1h1.5a.5.5 0 0 0 0-1H15v-1h1.5a.5.5 0 0 0 .5-.5V15a2 2 0 0 1 2 2v4a2 2 0 0 1-2 2h-4a2 2 0 0 1-2-2v-4a2 2 0 0 1 2-2Z" />
+    );
+  }
+
+  if (mimeType.startsWith("text/")) {
```
Comment (CTNicholas) [inline packages/liveblocks-react-ui/src/components/internal/Attachment.tsx:59]:
```suggestion
  if (mimeType.startsWith("text/") || mimeType.startsWith("application/")) {
```

I prefer your text file icon to the blank file icon. Looking at the [list of mimetypes](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types/Common_types), most of the common `application` mimetypes are actually documents or text files, (e.,g. MS word, rtf, ebook, JSON).

Alternatively, just show the few above as text files, as especially MS word `.doc(x)` files are common.

File: packages/liveblocks-react-ui/src/components/internal/Attachment.tsx [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+      <path d="M12 15.71a1 1 0 0 1 1.49-.872l4.96 2.79a1 1 0 0 1 0 1.744l-4.96 2.79A1 1 0 0 1 12 21.29v-5.58Z" />
+    );
+  }
+
+  if (mimeType.startsWith("audio/")) {
+    return (
+      <path d="M15 15a.5.5 0 0 0-.5.5v7a.5.5 0 0 0 1 0v-7a.5.5 0 0 0-.5-.5Zm-2.5 2.5a.5.5 0 0 1 1 0v3a.5.5 0 0 1-1 0v-3Zm-2 1a.5.5 0 0 1 1 0v1a.5.5 0 0 1-1 0v-1Zm6-1a.5.5 0 0 1 1 0v3a.5.5 0 0 1-1 0v-3ZM19 16a.5.5 0 0 0-.5.5v5a.5.5 0 0 0 1 0v-5a.5.5 0 0 0-.5-.5Z" />
+    );
+  }
+
```
Comment (CTNicholas) [inline packages/liveblocks-react-ui/src/components/internal/Attachment.tsx:82]: Could add a `startsWith("font/") && "application/vnd.ms-fontobject"` icon if we're being really picky!

**Review by CTNicholas — COMMENTED** [review]
Feel free to ignore any of these!

File: packages/liveblocks-react-ui/src/components/internal/Attachment.tsx (hunk 3 above)
Comment (marcbouchenoire) [reply packages/liveblocks-react-ui/src/components/internal/Attachment.tsx:82]: I'm adding them to the "text" icon list for now 👍

**Review by CTNicholas — COMMENTED** [review]
Left some things for me to improve here, when I get the chance.

Additionally:
- Mention new Lexical props for attachment
- Mention attachments in useCreateThread

**Review by CTNicholas — APPROVED** [review]
✅ Docs

**Review by FlowFlorent — APPROVED** [review]

File: packages/liveblocks-core/src/lib/autoRetry.ts [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+ * times. Between each attempt, it will inject a a backoff delay (in millis)
+ * from the given array. If the array contains fewer items then `maxTries`,
+ * then the last backoff number will be used indefinitely.
+ *
+ * If the last attempt is rejected too, the returned promise will fail too.
+ *
+ * @param promiseFn The promise factory to execute
+ * @param maxTries The number of total tries (must be >=1)
+ * @param backoff An array of timings to inject between each promise attempt
+ * @param throwError An optional function to not auto-retry on certain errors
```
Comment (nvie) [inline packages/liveblocks-core/src/lib/autoRetry.ts:15]: 👍

File: packages/liveblocks-react-ui/src/utils/use-latest.ts [hunk 5]
```diff
@@ -1,12 +1,18 @@
...
-  const ref: MutableRefObject<T> = useRef(value);
-
+/**
+ * Keeps a ref in sync with a given value that may or may not change on
+ * every render.
+ *
+ * The purpose of this hook is to return a stable ref that can be passed
+ * to a callback function so the callback can be registered but still can
+ * access the latest value at a later point in time.
+ */
```
Comment (nvie) [inline packages/liveblocks-react-ui/src/utils/use-latest.ts:11]: 🙏

File: packages/liveblocks-react-ui/src/components/Composer.tsx [hunk 6]
```diff
@@ -503,10 +662,33 @@ export const Composer = forwardRef(
...
-          onFocus={onFocus}
-          asChild
+          className={classNames(
+            "lb-root lb-composer lb-composer-form",
+            className
+          )}
+          dir={$.dir}
+          {...props}
+          ref={forwardedRef}
+          data-collapsed={isCollapsed ? "" : undefined}
```
[the remark this answers is not in the document]
Comment (marcbouchenoire) [reply packages/liveblocks-react-ui/src/components/Composer.tsx:672]: It's quite confusing but setting a data attribute to `true` will result in `data-attribute="true"`, setting it to `""` results in the expected `data-attribute` without any value 🤷

**Pull request merged**